File: demo-task-executor-scala/src/main/scala/demo/logger/LoggerFactory.scala
package demo.logger

import java.util.Objects

/**
  * Factory that builds a logger facade that writes messages to many [demo.logger.Logger] simultaneously.
  * This class IS thread-safe.
  */
class LoggerFactory(clazz: Class[_]) {

    val className = clazz.getPackage.getName + "." + clazz.getSimpleName.split("\\$").last
    val loggers = new collection.mutable.HashMap[Logger, Severity]()

    /**
      * Add a logger to the logging factory.
      *
      * @param logger   the logger.
      * @param severity the logging severity for the logger.
      */
    def add(logger: Logger, severity: Severity): LoggerFactory = {
        this.synchronized {
            loggers.put(Objects.requireNonNull(logger), Objects.requireNonNull(severity))
            return this
        }
    }

    /**
      * Builds a thread-safe logger facade from the added loggers.
      *
      * @return the thread-safe logger facade.
      */
    def build(): Logger = new AbstractLogger {
        override def log(severity: Severity, message: Any) {
            this.synchronized {
                for (logger <- loggers.keySet) {
                    if (SeverityConverter.toCode(severity) >= SeverityConverter.toCode(loggers.get(logger).get)) {
                        logger.log(severity, className + " - " + message)
                    }
                }
            }
        }
    }
}

File: demo-task-executor-java/src/main/java/demo/logger/LoggerFactory.java
package demo.logger;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Factory that builds a logger facade that writes messages to many {@link demo.logger.Logger} simultaneously.
 * This class IS thread-safe.
 */
final public class LoggerFactory {

    private final String className;
    private final Map<Logger, Severity> loggers;

    /**
     * Creates the logging factory for a client class.
     *
     * @param clazz the client clazz.
     */
    public LoggerFactory(Class<?> clazz) {
        this.className = Objects.requireNonNull(clazz).getName();
        this.loggers = new HashMap<>();
    }

    /**
     * Add a logger to the logging factory.
     *
     * @param logger   the logger.
     * @param severity the logging severity for the logger.
     */
    synchronized public LoggerFactory add(Logger logger, Severity severity) {
        loggers.put(Objects.requireNonNull(logger), Objects.requireNonNull(severity));
        return this;
    }

    /**
     * Builds a thread-safe logger facade from the added loggers.
     *
     * @return the thread-safe logger facade.
     */
    public Logger build() {
        return new AbstractLogger() {
            @Override
            synchronized public void log(Severity severity, Object message) {
                for (Logger logger : loggers.keySet()) {
                    if (severity.getValue() >= loggers.get(logger).getValue()) {
                        logger.log(severity, className + " - " + message);
                    }
                }
            }
        };
    }
}
